events/events.go

```diff
@@ -0,0 +1,104 @@
+package events
+
+import (
+	"context"
+	"time"
+
+	"github.com/reddit/baseplate.go/mqsend"
+
+	"github.com/apache/thrift/lib/go/thrift"
+)
+
+// Configuration values for the message queue.
+const (
+	// Max size in bytes for a single, serialized event.
+	MaxEventSize = 102400
+
+	// Max size of the events allowed in the message queue at one time.
+	MaxQueueSize = 10000
+
+	// Prefix added to the message queue name.
+	QueueNamePrefix = "events-"
+
+	// The default MaxPutTimeout to be used.
+	DefaultMaxPutTimeout = time.Millisecond * 50
+
+	// The default message queue name for v2 events.
+	DefaultV2Name = "v2"
+)
+
+// A Queue is an event queue.
+type Queue struct {
+	queue      mqsend.MessageQueue
+	maxTimeout time.Duration
+}
+
+// The Config used to initialize an event queue.
+type Config struct {
+	// The name of the message queue, should not contain the "events-" prefix.
+	//
+	// For v2 events, the default name (when passed in Name is empty) is "v2".
+	Name string
+
+	// The max timeout applied to Record function.
+	//
+	// If the passed in context object has a earlier deadline set,
+	// that deadline will be respected instead.
+	//
+	// If MaxPutTimeout <= 0, DefaultMaxPutTimeout will be used instead.
+	MaxPutTimeout time.Duration
+}
+
+// V2 initializes a new v2 event queue with default configurations.
+func V2() (*Queue, error) {
+	return V2WithConfig(Config{})
+}
+
+// V2WithConfig initializes a new v2 event queue.
+func V2WithConfig(cfg Config) (*Queue, error) {
+	name := cfg.Name
+	if name == "" {
+		name = DefaultV2Name
+	}
+	queue, err := mqsend.OpenMessageQueue(mqsend.MessageQueueConfig{
+		Name:           QueueNamePrefix + name,
+		MaxQueueSize:   MaxQueueSize,
+		MaxMessageSize: MaxEventSize,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	maxTimeout := cfg.MaxPutTimeout
+	if maxTimeout <= 0 {
+		maxTimeout = DefaultMaxPutTimeout
+	}
+
+	return &Queue{
+		queue:      queue,
+		maxTimeout: maxTimeout,
+	}, nil
+}
+
+// Close closes the event queue.
+//
+// After Close is called, all Put calls will return errors.
+func (q Queue) Close() error {
+	return q.queue.Close()
+}
+
+// Put serializes and puts an event into the event queue.
+func (q Queue) Put(ctx context.Context, event thrift.TStruct) error {
+	ctx, cancel := context.WithTimeout(ctx, q.maxTimeout)
+	defer cancel()
+
+	serializer := thrift.NewTSerializer()
+	serializer.Protocol = thrift.NewTJSONProtocol(serializer.Transport)
```

Comment:
🔕 Should we build the serializer on the `Queue` itself (or make it global) rather than rebuilding each time?

Reply:
`thrift.TSerializer` is not thread safe (we were bitten by it before :) )

It might make sense to use a resource pool for it, though.

Reply:
> thrift.TSerializer is not thread safe (we were bitten by it before :) )

What happened?

> It might make sense to use a resource pool for it, though.

I wonder if that might be a premature optimization, a resource pool is a more complicated to implement than just a global/class var and I'm not sure this actually causes problems.

Reply:
We tried to use a global serializer for everything and ended up with corrupt data everywhere.

This is actually the perfect use case of a resource pool. The main part we will save from using a resource pool is the initialization and gc time.

Reply:
Added resource pool and also tests. PTAL.